src/Database/Relations/BelongsTo.php

```diff
@@ -28,25 +28,77 @@ public function __construct(Builder $query, Model $child, $foreignKey, $ownerKey
      */
     public function add(Model $model, $sessionKey = null)
     {
+        /**
+         * @event model.relation.beforeAssociate
+         * Called before associating a relation to the model (only for BelongsTo/MorphTo relations)
+         *
+         * Example usage:
+         *
+         *     $model->bindEvent('model.relation.beforeAssociate', function (string $relationName, \October\Rain\Database\Model $relatedModel) use (\October\Rain\Database\Model $model) {
+         *         TODO: add example code
+         *     });
+         *
+         */
+        $this->parent->fireEvent('model.relation.beforeAssociate', [$this->relationName, $this->related]);
+
         if ($sessionKey === null) {
             $this->associate($model);
         }
         else {
             $this->child->bindDeferred($this->relationName, $model, $sessionKey);
```

Comment:
This brings to mind a question, should the events be fired if the relationship has been deferred or should they wait until the deferred relationship has been completed? @bennothommo any thoughts?

Reply:
@LukeTowers I updated these to fire in the `associate/dissociate` methods instead, which resolves this question I think...